### PR TITLE
Subclass MXCrypto to enable work-in-progress Rust sdk

### PR DESCRIPTION
**Pull request by Anderas**
To avoid long-running feature branch when implementing [Matrix Rust SDK](https://github.com/matrix-org/matrix-rust-sdk/tree/main/crates/matrix-sdk-crypto) aka Element-R, I'd like to start merging small changes into `develop` that are well isolated and cannot be shipped to production by accident. In particular all newly added code is hidden both behind a build flag as well as behind `DEBUG` compiler macro, so the new code cannot be accidentally used and shipped to production builds.

This particular PR sets up the skeleton for new rust-based implementation by overriding `MXCrypto` and all of its methods. This has both benefits and risks:
- no need to do any risky refactor across the rest of the application (for example by definining shared protocol)
- existing integration tests can still run unaltered, and validate that future Rust implementation works as expected
- potential risk: compiler does not ensure that each `MXCrypto`'s method is overriden, so if new public API was added to `MXCrypto`, this would not be automatically picked up by `MXCryptoV2`. This is a small risk, as the code is not meant for production yet, and when productionizing, more safety measures will be put in place.

Finally note this PR does not yet add any actual Rust library, merely sets the stage up.

### CODE REVIEW

File: MatrixSDK.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -7237,6 +7243,7 @@
 				MACOSX_DEPLOYMENT_TARGET = 10.10;
 				MTL_ENABLE_DEBUG_INFO = YES;
 				ONLY_ACTIVE_ARCH = YES;
+				OTHER_SWIFT_FLAGS = "-D DEBUG";
```
Comment (Anderas) [inline MatrixSDK.xcodeproj/project.pbxproj:7246]: Note the addition of this flag, otherwise `DEBUG` macro in Swift is incompatible with the one in objective-c

File: MatrixSDK/MXSDKOptions.h [hunk 2]
```diff
@@ -200,10 +200,22 @@ NS_ASSUME_NONNULL_BEGIN
  Enable sharing of session keys for an immediate historical context (e.g. last 10-20 messages)
  when inviting a new user to a room with shared history.
  
- @remark YES by default.
+ @remark NO by default.
```
Comment (Anderas) [inline MatrixSDK/MXSDKOptions.h:203]: Unrelated change, docs were stating incorrect default value

**Review by pixlwave — APPROVED** [review]
Looks good to me and the use of the build setting and compiler macro sounds sane. Two minor comments:
- Should we be concerned about other library users if they also have the `DEBUG` compiler macro set? I think the answer is no, but seems worth asking in case.
- The doc comments in the Swift file should probably use `///` instead of `/** */`

**Comment by Anderas** [comment]
>     * Should we be concerned about other library users if they also have the `DEBUG` compiler macro set? I think the answer is no, but seems worth asking in case.

Could you elaborate on what the worry is pls? To be clear, having `DEBUG` configuration by itself is not enough to use the new Crypto, one also has to set the build flag to true, the flag is just prevent submitting anything to production.

File: MatrixSDK/Data/MXRoom.m [hunk 3]
```diff
@@ -3643,7 +3643,7 @@ - (void)validateEncryptionStateConsistency
     if (!crypto)
     {
 #ifdef MX_CRYPTO
-        MXLogFailure(@"[MXRoom] checkEncryptionState: Crypto module is not present");
+        MXLogError(@"[MXRoom] checkEncryptionState: Crypto module is not present");
```
Comment (Anderas) [inline MatrixSDK/Data/MXRoom.m:3646]: There are some tests where these conditions actually occur. I do not yet know whether this is expected or not, either way the assertion will stop the tests. Temporarily switching to logging errors instead of failures and will investigate the failing tests individually.

**Comment by pixlwave** [comment]
> Could you elaborate on what the worry is pls? To be clear, having `DEBUG` configuration by itself is not enough to use the new Crypto, one also has to set the build flag to true, the flag is just prevent submitting anything to production.

It was only if using `DEBUG` meant further PRs might introduce small changes that were behind the `DEBUG` configuration but not also the build flag. Seemed unlikely but just asking in case :)

**Review by manuroe — APPROVED** [review]
Looks smart to me :D

**Comment by Anderas** [comment]
> > Could you elaborate on what the worry is pls? To be clear, having `DEBUG` configuration by itself is not enough to use the new Crypto, one also has to set the build flag to true, the flag is just prevent submitting anything to production.
> 
> It was only if using `DEBUG` meant further PRs might introduce small changes that were behind the `DEBUG` configuration but not also the build flag. Seemed unlikely but asked in case. :)

I see. Yea I do not currently anticipate such change, but will have to be careful about this.

I would actually like to get rid of the feature flag and replace it with compiler macro altogether (e.g. `MX_CRYPTO_V2`). The trouble here is that it needs to be usable from both obj-c and Swift and macros in each aren't currently mutally interchangable. The only reason DEBUG works this way is because it is manually set in build settings of the project. I may try to find another solution for this in the next PR, where the creation of the `MXCryptoV2` happens in some swift extension of `MXCrypto.m`, which means they could refer to the same Swift-only compiler macro.
